### PR TITLE
[persist] Add a structured key lower

### PR DESCRIPTION
**Pull request by bkirwi**
The "key lower" is a lower bound on the range of keys that are contained in a part. This adds a second key lower to each part - which is stored as structured data, and contains a lower bound in structured-data order.

Like other stats or lower bounds, we'd like to limit the amount of space the data takes. Here, we adopt the same strategy as we use for stats - trimming the data in its proto-encoded form so we can give precise size estimates.

This is tricky in two ways:
- We need to be careful to produce arrays that are still valid after being trimmed.
- We need to take advantage of the fact that a truncated struct array can still be compared to the original.

So I've added a debug assert and extended the proptest to check on that!

### Motivation

Support for #24830.

In particular, the key lower is very useful for the fast-path peek optimization when the data is not sorted: it means we often only need to fetch a handful of parts instead of all of them. I expect this to be particularly important during the migration, when we can't rely on the ordering of many runs to match our "desired" order.

### Tips for reviewer

Nightlies: https://buildkite.com/materialize/nightly/builds/9453

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] If this PR includes major [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note), I have pinged the relevant PM to schedule a changelog post.

### CODE REVIEW

File: src/persist-types/src/arrow.rs [hunk 1]
```diff
@@ -417,3 +423,115 @@ impl<'a> PartialEq for ArrayIdx<'a> {
...
+        self.ord.at(self.index)
+    }
+
+    /// Convert to an array-data proto, respecting a maximum data size. The resulting proto will
+    /// decode to a single-row array, such that `ArrayBound::new(decoded, 0).get() <= self.get()`,
+    /// which makes it suitable as a lower bound.
+    pub fn to_proto_lower(&self, max_len: usize) -> Option<ProtoArrayData> {
+        let array_data = match self.raw.len() {
+            0 => return None,
+            1 => self.raw.to_data(),
```
Comment (danhhz) [inline src/persist-types/src/arrow.rs:456]: Does this do the right thing if the input is a slice of some larger array? I.e. should we just always fall into the case below?

File: src/persist-types/src/arrow.rs [hunk 2]
```diff
@@ -301,6 +302,11 @@ impl ArrayOrd {
 
 /// A struct representing a particular entry in a particular array. Most useful for its `Ord`
 /// implementation, which can compare entire rows across similarly-typed arrays.
+///
+/// It is an error to compare indices from arrays with different types, with one exception:
```
Comment (danhhz) [inline src/persist-types/src/arrow.rs:306]: Might be nice to have a couple unit tests explicitly cover this case and also two structs where neither is a prefix of the other.

**Review by danhhz — APPROVED** [review]
LGTM!

There are a few spots here where things are getting pretty far indented. Would be nice to do a pass to fix those up, but we don't have any sort of style guide for it, so go ahead and skip if that's not something you're bought into.

File: src/persist-client/src/iter.rs [hunk 3]
```diff
@@ -481,6 +483,21 @@ where
...
                         part_desc: desc.clone(),
+                        structured_lower: match &part {
+                            BatchPart::Hollow(part) => {
+                                // Attempt to obtain the key lower, failing open if we can't obtain
+                                // it for any reason.
+                                part.structured_key_lower
+                                    .as_ref()
+                                    .and_then(|lazy| lazy.decode().ok())
+                                    .and_then(|data| ArrayData::from_proto(data).ok())
+                                    .and_then(|data| {
```
[the remark this answers is not in the document]
Comment (bkirwi) [reply src/persist-client/src/iter.rs:494]: Definitely is - thanks!

File: src/persist-types/src/arrow.rs (hunk 1 above)
Comment (bkirwi) [reply src/persist-types/src/arrow.rs:456]: Yeah, I think that's best. (The answer is some flavour of "not always", and it seems enough that it's worth the very marginal cost to have a simpler one.)

Comment (bkirwi) [reply src/persist-types/src/arrow.rs]: Done! Definitely feels more straightforward.

File: src/persist-types/src/arrow.rs (hunk 2 above)
Comment (bkirwi) [reply src/persist-types/src/arrow.rs:306]: Done!

**Comment by bkirwi** [comment]
> There are a few spots here where things are getting pretty far indented. Would be nice to do a pass to fix those up, but we don't have any sort of style guide for it, so go ahead and skip if that's not something you're bought into.

Took care of a few of these!

**Pull request merged**